tests/test_base.py

```diff
@@ -636,3 +655,189 @@ def test_pack_header_for_user(self, app, user_class):
+        verifies that the encrypted password matches the supplied scheme.
+        """
+        stock_config = app.config.copy()
+
+        _pwd = dirname(dirname(abspath(__file__)))
+        here = _pwd + '/flask_praetorian/templates'
+        tmpl_file = here + '/registration_email.html'
+        with open(tmpl_file) as _template:
+            tmpl = Template(_template.read())
+
+        # generate (w/o sending) a notification email and match it to our own
+        app.config['TESTING'] = True
+        app.config['PRAETORIAN_EMAIL_TEMPLATE'] = tmpl_file
+        app.config['PRAETORIAN_CONFIRMATION_ENDPOINT'] = 'unprotected'
+
+        default_guard = Praetorian(app, user_class)
+
+        # create our default test user
+        the_dude = user_class(
+            username='TheDude',
+            email='the@dude.com',
+            password=default_guard.hash_password('abides'),
+        )
+        db.session.add(the_dude)
+        db.session.commit()
+
+        with app.mail.record_messages() as outbox:
+            notify = default_guard.send_registration_email(user=the_dude)
+            the_dude.token = notify.token
+
+            # test our own interpretation and what we got back from flask_mail
+            assert tmpl.render(the_dude.__dict__).strip() in notify.message
+            assert notify.message == outbox[0].body
+
+            # test we got an expected confirmation URI
+            assert notify.confirmation_uri.endswith('unprotected/{}'.
+                                                    format(notify.token))
+
+            # test for no errors
+            assert not notify.errors
+
+        # test our token is good
+        assert default_guard.validate_confirmation(notify.token)
+
+        # test a bad token is treated as bad
+        with pytest.raises(PraetorianError):
+            default_guard.validate_confirmation('not a token')
+
+        # put away your toys
+        app.config = stock_config
+        db.session.delete(the_dude)
+        db.session.commit()
+
+    def test_validate_and_update(self, app, user_class, db):
+        """
+        This test verifies that Praetorian encrypts passwords using the scheme
+        specified by the HASH_SCHEME setting. If no scheme is supplied, the
+        test verifies that the default scheme is used. Otherwise, the test
+        verifies that the encrypted password matches the supplied scheme.
+        """
+        stock_config = app.config.copy()
+
+        default_guard = Praetorian(app, user_class)
+        pbkdf2_sha512_password = default_guard.hash_password('pbkdf2_sha512')
+
+        # create our default test user
+        the_dude = user_class(
+            username='TheDude',
+            email='the@dude.com',
+            password=pbkdf2_sha512_password,
+        )
+        db.session.add(the_dude)
+        db.session.commit()
+
+        """
+        Test the current password is hashed with PRAETORIAN_HASH_SCHEME
```

Comment:
probably better to break these out into multiple unit tests

Reply:
why so?

Reply:
First, you can document the specific behavior in each unit test so the comments aren't necessary or can go in the docstring. Second, there's less chance of side-effects from previous logical units in subsequent units. Finally, it makes it easier to deal with functionality changes later because of better isolation.